### PR TITLE
AP_HAL_Linux: Add heap support

### PR DESCRIPTION
**Pull request by WickedShell**
Add support for creating heaps. This is needed to let scripting run on Linux boards. The implementation was just copied from SITL. Still needs to be run on a real board.

### CODE REVIEW

**Comment by WickedShell** [comment]
This was tested/runs as expected on the PocketPilot/PocketBeagle. Thanks to @mirkix for the assistance with that!

**Comment by lucasdemarchi** [comment]
*heap*? is this name correct? Every malloc'ed memory in Linux comes from heap - I'm failing to see the connection to running scripts

File: libraries/AP_HAL_Linux/Util.h [hunk 1]
```diff
@@ -113,6 +119,18 @@ class Util : public AP_HAL::Util {
...
+
+#ifdef ENABLE_HEAP
+    struct heap_allocation_header {
+        size_t allocation_size; // size of allocated block, not including this header
+    };
+
+    struct heap {
+      size_t scripting_max_heap_size;
+      size_t current_heap_usage;
+    };
```
Comment (lucasdemarchi) [inline libraries/AP_HAL_Linux/Util.h]: this seems to be in the wrong abstraction layer...  If you need a "allocate_heap()" to be compatible with chibios/microcontrollers, then the HAL function should just allocate the memory in the right place and know nothing about scripting. Let the scripting part in AP_Script.

**Comment by WickedShell** [comment]
@lucasdemarchi Yeah, it shouldn't say scripting in the HAL layer you are absolutely correct. Scripting is currently the only user of the heap. The HAL method is intended to provide an abstracted interface to a managed/consistent heap. Other then fixing the internal naming (both here and in SITL) though, I do believe heap is an appropriate name.

**Comment by lucasdemarchi** [comment]
If you are returning from malloc, then it may make sense. Looking at the Chibios implementation it seems it's more similar to what it should be in Linux. Except that you don't need any fallback to allocate from elsewhere... If malloc fails, it fails.

**Comment by WickedShell** [comment]
@lucasdemarchi ChibiOS provides the utilities to create a managed pool of memory, which has been wrapped here. The target of this is to create a single contiguous pool of memory from which we can allocate/reallocate/free from. I don't know of any built in Linux library tools that provide this feature. I'm happy to pull the scripting bits out of naming, that shouldn't have been done, but I'm not sure what else to do about this?

**Comment by lucasdemarchi** [comment]
Why do you want to wrap exactly what malloc does on Linux? There are some reasons for that, but I don't think they apply here. On Chibios, afaiu from the code, it does that when malloc fails and try to allocate from elsewhere.

**Comment by WickedShell** [comment]
@lucasdemarchi The main point is that we want a constrained maximum allocation size, which we needed tracking for, which was more of a problem with using realloc(). We also wanted to provide the same experience across all boards so that you could run scripts in similar sized environments. There can be a bit of an argument that HAL_Linux doesn't need to restrict itself to a smaller memory pool, but it does mean you are interacting much more with the system's memory pool which we were trying to avoid.

I actually originally provided realloc() on chibios and just used that everywhere, but this meant that scripting got to consume all your memory, which could cause problems in flight, and we did want to try and pen the entire memory footprint of scripting to a single region. (In the future we'd actually like to turn on memory protection, which also should benefit from being a single lump area that is protected).

**Comment by WickedShell** [comment]
@lucasdemarchi Any update on the last comments from me? The overall desire is to keep everything behaving the same, and I'd be really worried about what happens if we start allocating into swap because of a runaway script.

**Comment by WickedShell** [comment]
@lucasdemarchi following the call from last week I brought this in, as it unblocks scripting on Linux, and given the above concerns on swap. I'm happy to continue the discussion/API here or on an issue though.

**Comment by lucasdemarchi** [comment]
I don't like this for several reasons:

1) we should really stop adding things to the Util class. It became the kitchen sink. 

2) "realloc" has a different meaning and semantics than the one provided here. At least a better name could be used to avoid misunderstandings

3) the abstraction is in the wrong place, making you have to copy and paste the implementation for different HALs, like was done from sitl -> Linux.

I'm on vacations with limited access to a computer so I can't fix what I think is wrong. But I think we should raise the bar a little bit... It should not be just "does it work?", because otherwise we spend a lot of time cleaning up things later.

**Pull request merged**